### ISSUE TITLE
Cannot open PDF for annotation due to reverse proxy issues

### ISSUE THREAD

**Issue by tpluscode** [inception-project/inception#4630]
**Describe the bug**

When annotation screen is opened, a message `An error occurred while loading the PDF` is shown

**To Reproduce**
1. Go to 'Administration/Documents'
2. Upload a PDF with parentheses in name
3. Go to annotate and select that file

**Please complete the following information:**
 - Version and build ID: [31.2 (2024-03-05 17:28:17, build 7c2c5b30)
 - OS: OS X Catalina
 - Browser: Brave 1.63.169 Chromium: 122.0.6261.111 (Official Build) (x86_64)

**Comment by reckart** [comment]
I have imported a file called `(lolo) lala.pdf`. The import worked and I could also open it in the PDF editor (using 32.0-SNAPSHOT). Do you have a stack trace?

**Comment by tpluscode** [comment]
Hm, I tried on the demo instance but did not reproduce. 

In the console I find these

```
Uncaught (in promise) TypeError: Failed to fetch
    at PQ.loadVisualModel (PDFAnnoPage.ts:154:12)
    at qB (pdfanno.ts:433:14)
    at xs (pdfanno.ts:94:3)
    at ge.init (PdfAnnotationEditor.ts:44:12)
    at fe.getOrInitialize (PdfAnnotationEditorFactory.ts:33:18)
    at ExternalEditor-ver-A79189520F29F3E484C12E0F19CB3501.js:1:3132
    at new Promise (<anonymous>)
    at c.initEditor (ExternalEditor-ver-A79189520F29F3E484C12E0F19CB3501.js:1:3041)
    at ExternalEditor-ver-A79189520F29F3E484C12E0F19CB3501.js:1:1174

Refused to connect to 'http://annotate.wikibus.org/p/wikibus-brochures/annotate?35-1.0-centerArea-editor-vis' because it violates the following Content Security Policy directive: "connect-src 'self'".

Refused to connect to 'http://annotate.wikibus.org/p/wikibus-brochures/annotate?35-1.0-centerArea-editor-vis' because it violates the document's Content Security Policy.
```

Looks like the file name was a red herring because now I'm getting this from various PDFs 🤔

**Comment by tpluscode** [comment]
Oh, wait, why are these http without the `s`...

**Comment by reckart** [comment]
Are you running behind a reverse proxy? Maybe it does not properly forward the external protocol?

Cf. https://inception-project.github.io/releases/31.3/docs/admin-guide.html#sect_reverse_proxy

**Comment by tpluscode** [comment]
Yes, I am. In my config:

```
wicket.core.csrf.accepted-origins[0]=annotate.wikibus.org
wicket.core.csrf.accepted-origins[1]=https://annotate.wikibus.org
```

I also notice that all navigation in the UI goes through a HTTP redirect first 🤔. The app only really works when I have both port 80 and port 443 exposed

**Comment by tpluscode** [comment]
![image](https://github.com/inception-project/inception/assets/588128/acf4b6ae-8a2d-4e56-9099-2435a7b96521)

**Comment by reckart** [comment]
Check your reverse proxy configuration. The reverse proxy has to tell INCEpTION that you are externally accessing it using https, typically using the request header `X-Forwarded-Proto`. Check the reverse proxy guide linked in my other comment.

**Comment by tpluscode** [comment]
I am hosting on clout66. It should be set up correctly and I have never had similar issue before. Here's the relevant portion of nginx config


```
    server {
        listen                  443 ssl http2;

        server_name             annotate.wikibus.org;

        client_max_body_size    250m;

        # redirect on errors
        error_page              500 502 504 /50x.html;

        # handle error redirect
        location = /50x.html {
            proxy_pass http://s3.amazonaws.com/placeholders.cloud66.com/nginx/50x.html;
        }

        location / {
            # Next three lines implement websocket support
            proxy_http_version 1.1;
            proxy_set_header Upgrade $http_upgrade;
            proxy_set_header Connection $connection_upgrade;
            proxy_set_header        X-Forwarded-For $proxy_add_x_forwarded_for;
            proxy_set_header        Host $http_host;
            proxy_set_header        X-Forwarded-Proto https;
            proxy_redirect          off;
            proxy_pass              http://inception;
            break;
        }
    }
```

**Comment by reckart** [comment]
You could try adding `server.forward-headers-strategy=NATIVE` to your `settings.properties`.

**Comment by tpluscode** [comment]
No change

**Comment by tpluscode** [comment]
The strange part is that I had no issues other than with parens in file names, or so I thought. And now I cannot open any files for annotation 😞

**Comment by reckart** [comment]
Very strange. Any recent changes to the overall setup?

**Comment by reckart** [comment]
When I try to access your instance, I see a bunch of errors in the JS console that I do not see when I try accessing one of our own instances, e.g. 

```
login.html:1 Refused to load the script 'https://annotate.wikibus.org/wicket/resource/org.apache.wicket.resource.JQueryResourceReference/jquery/jquery-3.7.1-ver-962A4A135B116C30221466C5E4F0E7A1.js' because it violates the following Content Security Policy directive: "script-src 'none'". Note that 'script-src-elem' was not explicitly set, so 'script-src' is used as a fallback.
```

**Comment by reckart** [comment]
At least for the login page of your site, I don't get a 302 redirect.

**Comment by tpluscode** [comment]
On initial load, no, there is indeed no redirect. But log in + log out also goes over HTTP resource.

For annotate, I think the actual problematic URL is this: `https://annotate.wikibus.org/ui/view/pdfeditor2/viewer.html?pdf=http://annotate.wikibus.org/p/test/annotate?22-1.0-centerArea-editor-vis&vmodel=http://annotate.wikibus.org/p/test/annotate?22-1.1-centerArea-editor-vis`. That is where the blocked resource is found. I see also a bunch of similar `/annotate?22-1...` requests but all of those are HTTPS

**Comment by reckart** [comment]
Most URLs are constructed relative by INCEpTION, so the browser just uses the same protocol it used to access the page. However, there are a few places where INCEpTION generates absolute URLs. And this only works if it is aware of the protocol used in the browser. When running behind a reverse proxy, this information is typically provided by the `X-Forwarded-XXX` headers. 

Unfortunately, I don't know what part of frameworks that INCEpTION uses handles these headers...

You might consider turning logging up to 11 and take a deep dive into the messages that are generated... add these to your `settings.properties` and brace yourself...

```
logging.level.org.springframework=TRACE
logging.level.org.apache.wicket=TRACE
logging.level.org.apache.catalina=TRACE
```

**Comment by reckart** [comment]
https://github.com/inception-project/inception/blob/inception-31.3/inception/inception-pdf-editor2/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor2/view/PdfDocumentIFrameView.java#L160-L175

**Comment by tpluscode** [comment]
Thanks. Off wicket mailing list then...

**Comment by reckart** [comment]
This could be in Wicket but it could also be in Spring.

**Comment by tpluscode** [comment]
Two behemoths. Maybe they will offer same advice. I am really clueless as to why this started happening out of the blue

**Comment by reckart** [comment]
Do you see in the logs something related to the incoming request to the annotation page, maybe the headers that are sent by the reverse proxy to the application?

**Comment by reckart** [comment]
@tpluscode For example, when you look at the logs, what do you see here? a `https` or a `http` URI?

```
CsrfPreventionRequestCycleListener - Request Source URI: ...
```

**Comment by tpluscode** [comment]
Here they are

> CsrfPreventionRequestCycleListener - Request Source URI: https://annotate.wikibus.org
> OriginResourceIsolationPolicy - Origin https://annotate.wikibus.org matched whitelisted origin annotate.wikibus.org, request accepted
> CsrfPreventionRequestCycleListener - CSRF Origin https://annotate.wikibus.org was whitelisted, allowed for page de.tudarmstadt.ukp.clarin.webanno.ui.annotation.AnnotationPage
> DefaultHandshakeHandler - Processing request http://annotate.wikibus.org/ws with headers=[upgrade:"websocket", connection:"Upgrade", x-forwarded-for:"31.60.52.3", host:"annotate.wikibus.org", x-forwarded-proto:"https", pragma:"no-cache", cache-control:"no-cache", origin:"https://annotate.wikibus.org", sec-websocket-version:"13", user-agent:"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36", accept-encoding:"gzip, deflate, br", accept-language:"en-GB,en-US;q=0.9,en;q=0.8,pl;q=0.7", cookie:"JSESSIONID=xyz", sec-gpc:"1", sec-websocket-key:"base64", sec-websocket-extensions:"permessage-deflate; client_max_window_bits"]

**Comment by reckart** [comment]
Interesting - most places say `https` but the request still went against `http`.

I am building an info widget which might help us debug this further....

**Comment by reckart** [comment]
@tpluscode You can build the PR yourself https://github.com/inception-project/inception/pull/4646

... or wait until the CI build is done and try the JAR from there: https://build.ukp.informatik.tu-darmstadt.de/job/inception-github/job/PR-4646/

After logging in as admin, go to the "Administration" link from the top bar.

<img width="1462" alt="Screenshot 2024-03-18 at 22 19 40" src="https://github.com/inception-project/inception/assets/1410238/774be082-7714-4be5-ad0c-59f8b79c7ea8">

**Comment by reckart** [comment]
Nope docker images only for releases.

**Comment by reckart** [comment]
You can check out the repo and build:

```
$ mvn clean install -DskipTests -T4
$ cd inception/inception-docker
$ mvn -Pdocker clean docker:build
```

Cross building from Apple Silicon to x86 without pushing unfortunately is currently not available... but if you have a docker repo available to push to, you can e.g.

```
$ mvn -Pdocker-buildx clean package -Ddocker.image.name="my.docker.repo/inception-project/inception"
```

**Comment by reckart** [comment]
JAR available at: https://build.ukp.informatik.tu-darmstadt.de/job/inception-github/job/PR-4646/lastSuccessfulBuild/artifact/de/tudarmstadt/ukp/inception/app/inception-app-webapp/31.4-SNAPSHOT/

**Comment by reckart** [comment]
I have pushed an image here:

 https://github.com/inception-project/inception/pkgs/container/inception-snapshots/192721365?tag=32.0-SNAPSHOT

Note that if you upgrade your installation to that image and later want to downgrade back to 31.x release version,
there *might* be problems. So if you want to try it, I would advise doing so using a fresh data folder / database.
If you upgrade your existing system using this image, I would strongly recommend you make sure you have a backup
that you can go back to.

 https://github.com/inception-project/inception/blob/main/inception/inception-doc/src/main/resources/META-INF/asciidoc/admin-guide/upgrade_backup.adoc

**Comment by tpluscode** [comment]
Wow, thanks for this. I will try this later today

**Comment by tpluscode** [comment]
![image](https://github.com/inception-project/inception/assets/588128/94b424aa-61ad-48fa-b2cd-626555ecd203)

**Comment by reckart** [comment]
Can you tell your reverse proxy to also generate the other `X-Forwarded-XXX` headers from my example above?

**Comment by tpluscode** [comment]
Will see what I can do. Thank you

**Comment by reckart** [comment]
@tpluscode btw. I think I found the relevant piece of code in Wicket that is bugging us here. It has an extensive JavaDoc comment.

https://github.com/apache/wicket/blob/master/wicket-core/src/main/java/org/apache/wicket/protocol/http/servlet/XForwardedRequestWrapperFactory.java

I have not studied it in-depth, but I believe the `X-Forwarded-Proto` may not be considered unless there is also a `X-Forwarded-For`... and that one is only considered if it matches a  set of internal IPs which defaults to private IP ranges.

**Comment by tpluscode** [comment]
The UI does not show `X-Forwarded-For` but it is configured using the `$proxy_add_x_forwarded_for` variable. Wicket says it should trust `10/8` IP range by default. My internal IP is `10.0.0.2`. I also tried adding the other headers explicitly but without luck

**Comment by reckart** [comment]
In your earlier post, we saw this:

`DefaultHandshakeHandler - Processing request http://annotate.wikibus.org/ws with headers=[upgrade:"websocket", connection:"Upgrade", `**`"x-forwarded-for:"31.60.52.3"`**`, host:"annotate.wikibus.org", x-forwarded-proto:"https", pragma:"no-cache", cache-control:"no-cache", origin:"https://annotate.wikibus.org", sec-websocket-version:"13", user-agent:"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36", accept-encoding:"gzip, deflate, br", accept-language:"en-GB,en-US;q=0.9,en;q=0.8,pl;q=0.7", cookie:"JSESSIONID=xyz", sec-gpc:"1", sec-websocket-key:"base64", sec-websocket-extensions:"permessage-deflate; client_max_window_bits"]`

Here, we can see the `x-forwarded-for` header as well as the `x-forwarded-proto` header. This comes from the WebSocket part of the application that is handled by Spring.

We can also see it in the nginx configuration you have posted.

Yet, when I access the request headers on the administration page, it appears the header is not found. There, I ask Wicket for the request, then fetch the underlying container request (which should come from Tomcat). 

Hm, looking at my code, I see that I explicitly list the bunch of `x-forwarded-...` headers to look for but forgot the `x-forwarded-for` header - so we won't currently see that on the admin page. I'll update that code to list any headers starting with `x-forwarded-` and check on my working system again.

**Comment by reckart** [comment]
When you add the other headers in your reverse proxy, do they appear on the admin page?

**Comment by tpluscode** [comment]
Yes, they did appear

**Comment by reckart** [comment]
So I did some more research on these headers and reverse proxy support in spring.

* **X-Forwarded-For**:  header field is a common method for identifying the originating IP address of a client connecting to a web server through an HTTP proxy
* **X-Forwarded-Proto**: header is a de-facto standard header for identifying the protocol (HTTP or HTTPS) that a client used to connect to your proxy or load balancer.
* **X-Forwarded-Scheme**: Alternative to X-Forwarded-Proto with the same function
* **X-Forwarded-Host**: The original host requested by the client in the Host HTTP request header.
* **X-Forwarded-Port**: request header helps you identify the destination port that the client used to connect to the load balancer.
* **X-Forwarded-Server** is the hostname of the proxy server. It gets overwritten by each proxy, which is involved in the communication, with the current proxy's hostname.
* **X-Forwarded-By** is the proxy IP address.

Additionally, there are a few Spring Boot settings:

* `server.forward-headers-strategy` - If the proxy adds the commonly used X-Forwarded-For and X-Forwarded-Proto headers, setting server.forward-headers-strategy to `NATIVE` is enough to support those. With this option, the Web servers themselves natively support this feature; you can check their specific documentation to learn about specific behavior.
* `server.tomcat.redirect-context-root` - If you are using Tomcat and terminating SSL at the proxy, this should be set to `false`. This allows the X-Forwarded-Proto header to be honored before any redirects are performed.
* `server.tomcat.remoteip.remote-ip-header`, `server.tomcat.remoteip.protocol-header`, `server.tomcat.remoteip.internal-proxies` can be used to override various settings

*More info: https://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#howto.webserver.use-behind-a-proxy-server*

So my reverse proxy provides these headers:

* `x-forwarded-host`
* `x-forwarded-port`
* `x-forwarded-proto`
* `x-forwarded-scheme`
* `x-forwarded-server`

I believe that in particular `x-forwarded-host`, `x-forwarded-port` and `x-forwarded-proto` are required. In particular, I do not use `x-forwarded-for`.

Additionally, I am using `server.forward-headers-strategy=NATIVE` in the `settings.properties`.

I do not use `server.tomcat.redirect-context-root` - but maybe I should.

Could you try removing the `x-forwarded-for` header and instead add `x-forwarded-host`, `x-forwarded-port` and `x-forwarded-proto` and `server.forward-headers-strategy=NATIVE` in the `settings.properties` and report back?

**Comment by tpluscode** [comment]
I have already set the `NATIVE` strategy

Unfortunately, that and the headers is still not enough...

![image](https://github.com/inception-project/inception/assets/588128/3078eaa8-0411-496e-a6ea-004fb726d3d4)

`X-Forwarded-Host` is showing my IP

**Comment by reckart** [comment]
Try if you can make it so that x-forwarded-host instead passes on the originally requested host name.

The examples from my setup I gave you use Apache HTTPD as a reverse proxy.

Our documentation says for nginx, you should add 

```
proxy_set_header   Host             $host;
```

which appears to be missing from your configuration file.

See: https://inception-project.github.io/releases/31.3/docs/admin-guide.html#_nginx_as_reverse_proxy

**Comment by tpluscode** [comment]
YES! Finally! Thank you for your help. Your email suggestions worked.

In the end what did was the [`server.tomcat.remoteip.internal-proxies`](https://docs.spring.io/spring-boot/docs/current/reference/html/application-properties.html#application-properties.server.server.tomcat.remoteip.internal-proxies) setting. When I needed to add a pattern for the IP which appears as remote host. I copied the long default regular expression and added one more alternative but I suppose it would have sufficed to only set it to match my current host
